ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,381 @@
+            }
+            else
+            {
+                perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+            
+                UpgradeToExclusiveLock(s_entriesLock);
+                lockElevatedToExclusive = true;
+            
+                if (!FindAndUpdateEntryToLatest_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        vnodeFsidInode,
+                        cacheIndex,    // starting index
+                        indexFromHash, // stopping index
+                        vnodeVid,
+                        invalidateEntry,
+                        /* out */ rootHandle))
+                {
+                    ClearCacheAndInsertEntry_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        vnodeFsidInode,
+                        indexFromHash,
+                        vnodeVid,
+                        /* out */ rootHandle);
+                }
+            }
+        }
+        else
+        {
+            perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+            
+            UpgradeToExclusiveLock(s_entriesLock);
+            lockElevatedToExclusive = true;
+            
+            ClearCacheAndInsertEntry_ExclusiveLocked(
+                perfTracer,
+                cacheMissFallbackFunctionCounter,
+                cacheMissFallbackFunctionInnerLoopCounter,
+                vnode,
+                vnodeFsidInode,
+                indexFromHash,
+                vnodeVid,
+                /* out */ rootHandle);
+        }
+    }
+    
+    if (lockElevatedToExclusive)
+    {
+        RWLock_ReleaseExclusive(s_entriesLock);
+    }
+    else
+    {
+        RWLock_ReleaseShared(s_entriesLock);
+    }
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (perfTracer)
+    {
+        perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+    }
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+static inline void UpgradeToExclusiveLock(RWLock& lock)
+{
+    if (!RWLock_AcquireSharedToExclusive(lock))
+    {
+        RWLock_AcquireExclusive(lock);
+    }
+}
+
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode)
+{
+    uintptr_t vnodeAddress = reinterpret_cast<uintptr_t>(vnode);
+    return (vnodeAddress >> 3) % s_entriesCapacity;
```

Comment:
If Apple grows the `vnode` struct from 248 to 256 bytes, this will most likely suddenly become quite a bad hash function. (It's probably not great even now.) Not the absolute worst as `s_entriesCapacity` is usually not a power of 2, but I'd put a TODO here at minimum.

Reply:
> How did we come up with this algorithm for hashing?

After looking around online I found that a bit-shift hash is a common choice for hashing memory addresses.  There's been a good performance improvement with this initial hash function and I've not yet spent time trying other hash functions.  

@pmj brings up a good point.  @pmj as best I could tell vnodes are opaque from the perspective of our kext.  Do you know of any tricks we could use for determining their size or potential breaks to our hashing algorithm? 

@jeschu1, @pmj if you know of hashing algorithms that might be better suited for our scenario please let me know.

Reply:
> Do you know of any tricks we could use for determining their size or potential breaks to our hashing algorithm?

I'll have to check if we can get access to the settings for the zone allocator's vnode zone via public APIs. If not, we'd have to look at the low 12 bits of a bunch of vnode pointers, from those it should be pretty easy to infer, within some reasonable bounds anyway.

> @jeschu1, @pmj if you know of hashing algorithms that might be better suited for our scenario please let me know.

Not off the top of my head, but hashing pointer values is a pretty common thing to do, so there must be some very robust implementations out there. I seem to remember Google publishing some research results on hash tables a while back, that might include something on this topic.

Also highly relevant: table size strategy. A function might be a good hash function with power-of-2 table sizes, but bad for other sizes, or good for prime-sized tables but bad for others. So we might want to revisit the `2 * desiredvnodes` decision and amend that to e.g. `NextPrimeLargerThan(desiredvnodes)` or `RoundUpToPow2(desiredvnodes)`.

Whatever function/size strategy we decide to try, I think we should gather one or more data sets of real vnode addresses via a special version of the kext that logs them all, and then look at collision rates. Especially because there are definite patterns to vnode addresses, they are most certainly not random.

Reply:
@pmj thanks for the tips on table size and the reminder to get some vnode addresses for testing.

In the last commit I pushed I went with the approach of requiring the cache capacity to be a power of 2, which also allowed for switching from a `%` operation to faster a bit-wise `&`.

Reply:
> the reminder to get some vnode addresses for testing.

Filed #974 for this